Sparse CSR: Add tensor.resize_ and tensor.copy_

IvanYashchuk opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66401
* #62180
* #65606
* #64391
* #61536
* #61858
* #61857
* #63948
* #63511
* __->__ #63510

Sparse CSR matrix resizing behavior:
If we _increase the number of rows_ the number of specified elements in the matrix remains the same -> the size of col_indices, values doesn't change, the size of crow_indices becomes `rows+1`.
If we _decrease the number of rows_ the number of specified elements will be `min(nnz, rows*cols)` -> need to resize `crow_indices` to `rows+1` and set the last element to `min(nnz, rows*cols)`; decrease the size of col_indices and values to `min(nnz, rows*cols)`.
If we _increase the number of columns_ the number of specified elements in the matrix remains the same, the number of rows remains the same -> no need to resize anything, just set new sizes.
We _cannot decrease the number of columns_ because it would require recomputing `crow_indices`.


cc @nikitaved @pearu @cpuhrsch @IvanYashchuk

Differential Revision: [D31796680](https://our.internmc.facebook.com/intern/diff/D31796680)

IvanYashchuk committed:
Sparse CSR: Add tensor.resize_ and tensor.copy_

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add tensor.resize_ and tensor.copy_"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add tensor.resize_ and tensor.copy_"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add tensor.resize_ and tensor.copy_"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add tensor.resize_ and tensor.copy_"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add tensor.resize_ and tensor.copy_"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add tensor.resize_ and tensor.copy_"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add tensor.resize_ and tensor.copy_"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add tensor.resize_ and tensor.copy_"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add tensor.resize_ and tensor.copy_"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add tensor.resize_ and tensor.copy_"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add tensor.resize_ and tensor.copy_"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add tensor.resize_ and tensor.copy_"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add tensor.resize_ and tensor.copy_"




[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add tensor.resize_ and tensor.copy_"




cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add tensor.resize_ and tensor.copy_"




cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add tensor.resize_ and tensor.copy_"




cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add tensor.resize_ and tensor.copy_"




cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add tensor.resize_ and tensor.copy_"




cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add tensor.resize_ and tensor.copy_"




cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add tensor.resize_ and tensor.copy_"




cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add tensor.resize_ and tensor.copy_"




cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

cpuhrsch commented on test/test_sparse_csr.py:
```diff
@@ -134,6 +134,83 @@ def test_empty_errors(self, device, dtype):
+
+    @skipMeta
+    @dtypes(*get_all_dtypes())
+    def test_resize(self, device, dtype):
```
One feature of resize is that if the [memory shrinks the values are preserved](https://pytorch.org/docs/master/generated/torch.Tensor.resize_.html?highlight=resize). I think this should be true as well here since we use resize_ to change values, but it could be good to add a test just to make sure and in particular when resizing each dimension separately.

IvanYashchuk replied:
The behavior is similar here, yes because of `resize_` on strided tensors, but only for column indices and values. The last element of row indices is modified with the resize operation. Currently resizing to a smaller size and then back to the original size would not result in the same original matrix.

cpuhrsch replied:
Added another comment higher up in the code. What do you mean by ```the last element of row indices is modified with the resize operation```?

IvanYashchuk committed:
Update on "Sparse CSR: Add tensor.resize_ and tensor.copy_"




cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add tensor.resize_ and tensor.copy_"




cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add tensor.resize_ and tensor.copy_"




cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add tensor.resize_ and tensor.copy_"




cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add tensor.resize_ and tensor.copy_"




cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add tensor.resize_ and tensor.copy_"


Sparse CSR matrix resizing behavior:
If we _increase the number of rows_ the number of specified elements in the matrix remains the same -> the size of col_indices, values doesn't change, the size of crow_indices becomes `rows+1`.
If we _decrease the number of rows_ the number of specified elements will be `min(nnz, rows*cols)` -> need to resize `crow_indices` to `rows+1` and set the last element to `min(nnz, rows*cols)`; decrease the size of col_indices and values to `min(nnz, rows*cols)`.
If we _increase the number of columns_ the number of specified elements in the matrix remains the same, the number of rows remains the same -> no need to resize anything, just set new sizes.
We _cannot decrease the number of columns_ because it would require recomputing `crow_indices`.


cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add tensor.resize_ and tensor.copy_"


Sparse CSR matrix resizing behavior:
If we _increase the number of rows_ the number of specified elements in the matrix remains the same -> the size of col_indices, values doesn't change, the size of crow_indices becomes `rows+1`.
If we _decrease the number of rows_ the number of specified elements will be `min(nnz, rows*cols)` -> need to resize `crow_indices` to `rows+1` and set the last element to `min(nnz, rows*cols)`; decrease the size of col_indices and values to `min(nnz, rows*cols)`.
If we _increase the number of columns_ the number of specified elements in the matrix remains the same, the number of rows remains the same -> no need to resize anything, just set new sizes.
We _cannot decrease the number of columns_ because it would require recomputing `crow_indices`.


cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add tensor.resize_ and tensor.copy_"


Sparse CSR matrix resizing behavior:
If we _increase the number of rows_ the number of specified elements in the matrix remains the same -> the size of col_indices, values doesn't change, the size of crow_indices becomes `rows+1`.
If we _decrease the number of rows_ the number of specified elements will be `min(nnz, rows*cols)` -> need to resize `crow_indices` to `rows+1` and set the last element to `min(nnz, rows*cols)`; decrease the size of col_indices and values to `min(nnz, rows*cols)`.
If we _increase the number of columns_ the number of specified elements in the matrix remains the same, the number of rows remains the same -> no need to resize anything, just set new sizes.
We _cannot decrease the number of columns_ because it would require recomputing `crow_indices`.


cc nikitaved pearu cpuhrsch @IvanYashchuk

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Sparse CSR: Add tensor.resize_ and tensor.copy_"


Sparse CSR matrix resizing behavior:
If we _increase the number of rows_ the number of specified elements in the matrix remains the same -> the size of col_indices, values doesn't change, the size of crow_indices becomes `rows+1`.
If we _decrease the number of rows_ the number of specified elements will be `min(nnz, rows*cols)` -> need to resize `crow_indices` to `rows+1` and set the last element to `min(nnz, rows*cols)`; decrease the size of col_indices and values to `min(nnz, rows*cols)`.
If we _increase the number of columns_ the number of specified elements in the matrix remains the same, the number of rows remains the same -> no need to resize anything, just set new sizes.
We _cannot decrease the number of columns_ because it would require recomputing `crow_indices`.


cc nikitaved pearu cpuhrsch @IvanYashchuk

Differential Revision: [D31796680](https://our.internmc.facebook.com/intern/diff/D31796680)

[ghstack-poisoned]
